core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -324,6 +324,10 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     val app = try {
       load(appId)
      } catch {
+      case _: NoSuchElementException if this.conf.get(ON_DEMAND_ENABLED) =>
+        val name = Utils.nameForAppAndAttempt(appId, attemptId)
+        loadFromFallbackLocation(appId, attemptId,
+          RollingEventLogFilesWriter.EVENT_LOG_DIR_NAME_PREFIX + name)
```

Comment:
We should not assume it will be `RollingEventLogFilesWriter`, users dont need to be running with the default enabled.